### PR TITLE
Migrate Decoder to NNX

### PR DESCRIPTION
**Pull request by hsuan-lun-chiang**
# Description
Migrate the Transformer decoder layer into NNX.

Note: The following models are currently not supported:

- DeepSeek
- Gemma3
- Llama4

Support for these models will be added in a follow-up PR.

Strategy:
A `pure_nnx_decoder` flag is added to control whether NNX or Linen decoder shall be used.
Initial migration doesn't include the pipeline NNX support. 

# Tests
Conducted these tests. Details in the [GDoc file](https://docs.google.com/document/d/1NbUP3g5glgbC6bMyt44pwM_vQA1NR7U2rBUzfbTDwSs/edit?pli=1&resourcekey=0-9EUahtzL-hCycdu7l0grhQ&tab=t.htq5367h8au0)
1. Test with different model and compare with Linen training
2. Golden logits comparison
3. Inference
4. Checkpoint comparison (Including TreeStructure Comparison)
5. Sharding comparison

TODOs:
- NNX version of unit tests (future PRs)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/models.py [hunk 1]
```diff
@@ -85,8 +85,8 @@ def setup(self):
         config=cfg,
         mesh=self.mesh,
     )
-    self.vision_encoder = vision_encoder_as_linen(config=cfg, mesh=mesh) if cfg.use_multimodal else None
-    self.decoder = Decoder(config=cfg, mesh=mesh, quant=self.quant, model_mode=self.model_mode)
+    self.vision_encoder = VisionEncoder(config=cfg, mesh=mesh) if cfg.use_multimodal else None
```
Comment (cgarciae) [inline src/MaxText/layers/models.py]: You cannot assign NNX Modules to Linen Modules, you need to use the bridge API.

File: src/MaxText/layers/decoders.py [hunk 2]
```diff
@@ -64,915 +67,665 @@
...
+        name: str = "decoder_layer",
+    ):
+        self.config = config
+        self.mesh = mesh
+        self.model_mode = model_mode
+        self.quant = quant
+        
+        cfg = self.config
+
+        # 1. Pre-Self-Attention Norm
```
Comment (cgarciae) [inline src/MaxText/layers/decoders.py]: remove unnecessary comments

File: src/MaxText/layers/decoders.py [hunk 3]
```diff
@@ -64,915 +67,665 @@
...
-            image_masks=image_masks,
-        )
-      # TODO(hengtaoguo): Add support for other multimodal models such as Llama4, refactor if needed
+      nnx.update(layers, scanned_state)
+
+      return final_carry, None
+
+    def get_decoder_layers(self):
+      """Retrieves decoder layer classes based on config."""
+      match self.config.decoder_block:
```
Comment (cgarciae) [inline src/MaxText/layers/decoders.py]: wondering if all this logic could just be turned into a dictionary lookup?

File: src/MaxText/layers/decoders.py [hunk 4]
```diff
@@ -64,915 +67,665 @@
...
+          # layer_i is a slice of the vmapped layers, representing one layer
+          layer_out = layer(carry, *args, **kwargs)
+          # handle tuple returns (output, aux) -> just take output for carry
+          new_carry = layer_out[0] if isinstance(layer_out, tuple) else layer_out
+          _, new_layer_state = nnx.split(layer)
+          return new_carry, new_layer_state
+      
+      # Apply checkpointing to the step function
+      if policy is not None:
+          pure_layer_fn = jax.checkpoint(pure_layer_fn, policy=policy, prevent_cse=prevent_cse)
```
Comment (cgarciae) [inline src/MaxText/layers/decoders.py]: `pure_layer_fn` is undefined

File: src/MaxText/layers/decoders.py [hunk 5]
```diff
@@ -64,915 +67,665 @@
...
-          model_mode=self.model_mode,
+      prevent_cse = maxtext_utils.should_prevent_cse_in_remat(self.config)
+      graphdef, params, state = nnx.split(layers, nnx.Param, ...)  # state: the mutable state we carry (KV cache, RNGs, etc.)
+      def layer_fn(carry, layer_params):
+          layer = nnx.merge(graphdef, layer_params, other_state)
+          # layer_i is a slice of the vmapped layers, representing one layer
+          layer_out = layer(carry, *args, **kwargs)
+          # handle tuple returns (output, aux) -> just take output for carry
+          new_carry = layer_out[0] if isinstance(layer_out, tuple) else layer_out
+          _, new_layer_state = nnx.split(layer)
```
Comment (cgarciae) [inline src/MaxText/layers/decoders.py]:
```suggestion
          new_layer_state = nnx.state(layer)
```

File: src/MaxText/layers/models.py (hunk 1 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/models.py]: Thank you for pointing this out, I have updated the code. However, there's still some issues causing the PR tests failed and I'm working on it.

File: src/MaxText/layers/decoders.py (hunk 2 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/decoders.py]: Done.

File: src/MaxText/layers/decoders.py (hunk 3 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/decoders.py]: Thanks for the idea. I updated the function to retrieve the layer using a dictionary lookup instead of a match statement.

File: src/MaxText/layers/decoders.py (hunk 4 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/decoders.py]: Fixed.

File: src/MaxText/layers/decoders.py (hunk 5 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/decoders.py]: Done, thanks!

File: src/MaxText/layers/decoders.py [hunk 6]
```diff
@@ -61,932 +65,646 @@
...
-    # scan does not support kwargs in layer call, passing broadcast_args as positional arg
-    broadcast_args = (
-        decoder_segment_ids,
+        image_embeddings=None,
+        bidirectional_mask=None,
+        image_masks=None,
+    ):
+        cfg = self.config
+        
+        if isinstance(shared_embedding, nnx.Module):
```
Comment (RissyRan) [inline src/MaxText/layers/decoders.py]: This seems exactly same?

```

        if isinstance(shared_embedding, nnx.Module):
             y = shared_embedding(decoder_input_tokens.astype("int32"), model_mode=model_mode)
        else:
            y = shared_embedding(decoder_input_tokens.astype("int32"), model_mode=model_mode)
  ```

File: src/MaxText/layers/models.py [hunk 7]
```diff
@@ -315,10 +315,12 @@ def __init__(
...
     )
-    self.vision_encoder = VisionEncoder(config=cfg, mesh=mesh, rngs=rngs) if cfg.use_multimodal else None
-
-    decoder_linen = Decoder(config=cfg, mesh=mesh, quant=self.quant, model_mode=self.model_mode)
-    self.decoder = nnx_wrappers.ToNNX(decoder_linen, rngs=rngs)
+    self.vision_encoder = VisionEncoder(config=cfg, mesh=mesh) if cfg.use_multimodal else None
+    if rngs is not None:
+      self.decoder = Decoder(config=cfg, mesh=mesh, quant=self.quant, model_mode=self.model_mode, rngs=rngs)
+    else:
+      decoder_linen = Decoder(config=cfg, mesh=mesh, quant=self.quant, model_mode=self.model_mode)
```
Comment (RissyRan) [inline src/MaxText/layers/models.py]: Not needed for this branch?

File: src/MaxText/layers/decoders.py [hunk 8]
```diff
@@ -61,932 +65,646 @@
...
+            "previous_chunk": previous_chunk,
+            "page_state": page_state,
+            "slot": slot,
+            "attention_metadata": attention_metadata,
+        }
+
+        if cfg.decoder_block == DecoderBlockType.GEMMA3:
+            layer_kwargs["bidirectional_mask"] = bidirectional_mask
+
+        if cfg.scan_layers:
```
Comment (RissyRan) [inline src/MaxText/layers/decoders.py]: This will be very tricky! Some models are handled different, in this case, you will unfortunately need to test them. 

For scan training, at least:
1) DeepSeek (dense + moe layers)
2) Gemma3 with _apply_gemma3_scanned_blocks
3) Llama4 with inhomogeneous_layer_cycle_interval
...

You will also need to check unscanned conditions.

**Comment by hsuan-lun-chiang** [comment]
> Thanks for the change! Unfortunately this part involves all models using scan/unscan, and some patterns are different. Please be careful and include comprehensive test suit before the merge.

Sure, will do, and thanks for the heads-up!

File: src/MaxText/layers/models.py (hunk 7 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/models.py]: Yes, I've removed them.

File: src/MaxText/layers/decoders.py (hunk 6 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/decoders.py]: Thank you, I've removed the redundant code.

File: src/MaxText/layers/models.py [hunk 9]
```diff
@@ -362,13 +371,6 @@ def __init__(
     else:
       dummy_attention_metadata = None
 
-    self.decoder.lazy_init(
```
Comment (xibinliu) [inline src/MaxText/layers/models.py:359]: Should we keep this if pure_nnx_decoder is False?

File: src/MaxText/layers/models.py (hunk 1 above)
Comment (xibinliu) [reply src/MaxText/layers/models.py]: Code rebased and VisionEncoder is not changed.

File: src/MaxText/layers/decoders.py (hunk 8 above)
Comment (xibinliu) [reply src/MaxText/layers/decoders.py]: These will be supported with new PRs.

File: src/MaxText/layers/models.py (hunk 9 above)
Comment (hsuan-lun-chiang) [reply src/MaxText/layers/models.py:359]: Yes, thank you for pointing it out. Fixed.

**Review by bvandermoon — COMMENTED** [review]
@hsuan-lun-chiang thanks for the testing in the PR description. For the memory differences mentioned in the doc, could you collect before/after profiles so we can confirm the HLOs are the same?

**Comment by hsuan-lun-chiang** [comment]
> @hsuan-lun-chiang thanks for the testing in the PR description. For the memory differences mentioned in the doc, could you collect before/after profiles so we can confirm the HLOs are the same?

Hi @bvandermoon ,
Yes, we already have the profiles (for Gpt3, where the memory difference was most significant)
[Before (Linen)](https://xprof.corp.google.com/trace_viewer/hsuanlun-628791851930266594?hosts=t1v-n-653c47b4-w-0&host_index=0&trace_filter_config={})
[After (NNX)](https://xprof.corp.google.com/trace_viewer/hsuanlun-18082195794636712941?hosts=t1v-n-653c47b4-w-0&host_index=0&trace_filter_config={})
Thank you for checking the work!

**Comment by hsuan-lun-chiang** [comment]
> Why are these models being skipped for now? I think it's fine just want to understand if there is additional context
> 
> > Note: The following models are currently not supported:
> > DeepSeek
> > Gemma3
> > Llama4

No major blockers; these models require a different handling logic compared to other models (as @RissyRan commented earlier). We are addressing them in a separate workstream to ensure they are integrated correctly.

**Comment by bvandermoon** [comment]
> > @hsuan-lun-chiang thanks for the testing in the PR description. For the memory differences mentioned in the doc, could you collect before/after profiles so we can confirm the HLOs are the same?
> 
> Hi @bvandermoon , Yes, we already have the profiles (for Gpt3, where the memory difference was most significant) [Before (Linen)](https://xprof.corp.google.com/trace_viewer/hsuanlun-628791851930266594?hosts=t1v-n-653c47b4-w-0&host_index=0&trace_filter_config=%7B%7D) [After (NNX)](https://xprof.corp.google.com/trace_viewer/hsuanlun-4730774157194586265?hosts=t1v-n-653c47b4-w-0&host_index=0&trace_filter_config=%7B%7D) Thank you for checking the work!

Discussed offline, thanks @hsuan-lun-chiang. Looks like the profiles have different HLOs. Could you also collect profiles for the other ones in the first section of your doc? We can confirm if it's just GPT3 or something happening everywhere

**Comment by hsuan-lun-chiang** [comment]
> Discussed offline, thanks @hsuan-lun-chiang. Looks like the profiles have different HLOs. Could you also collect profiles for the other ones in the first section of your doc? We can confirm if it's just GPT3 or something happening everywhere

Sure! I'm collecting those profiles and will leave a comment to let you know where I documented them (in the testing page, section 1). Thank you!

**Comment by bvandermoon** [comment]
> > Discussed offline, thanks @hsuan-lun-chiang. Looks like the profiles have different HLOs. Could you also collect profiles for the other ones in the first section of your doc? We can confirm if it's just GPT3 or something happening everywhere
> 
> Sure! I'm collecting those profiles and will leave a comment to let you know where I documented them (in the testing page, section 1). Thank you!

Thanks @hsuan-lun-chiang. For Llama2-7b and Llama3.1-70B, the before/after memory usage are almost identical. But for some reason the HLOs are not identical. Can you confirm the before/after runs were run in the same env with the same JAX/libtu version, etc.?

**Comment by hsuan-lun-chiang** [comment]
> Thanks @hsuan-lun-chiang. For Llama2-7b and Llama3.1-70B, the before/after memory usage are almost identical. But for some reason the HLOs are not identical. Can you confirm the before/after runs were run in the same env with the same JAX/libtu version, etc.?

Thank you @bvandermoon for checking this. Yes, I ran the before/after tests by only switching branches like:
```
git checkout main
python3 -m MaxText.train ...
git checkout Migrate-Decoder-to-NNX
python3 -m MaxText.train ... nnx_enable=True pure_nnx_decoder=True
```

So both the main branch code and our code were tested in the same environment.
I will double check the implementation of both models to see if I can find anything.

File: src/maxtext/layers/multi_token_prediction.py [hunk 10]
```diff
@@ -108,22 +108,12 @@ def __init__(
         rngs=rngs,
     )
     # Use MODEL_MODE_TRAIN for initialization; runtime model_mode is passed dynamically.
-    mtp_transformer_layer = transformer_layer_module(
+    self.transformer_layer = transformer_layer_module(
```
Comment (cgarciae) [inline src/maxtext/layers/multi_token_prediction.py:111]: Keep the same `name` here to preserve the checkpoint structure e.g:

```python
setattr(self, f"mtp_{k}_transformer_layer", transformer_layer_module(...))
```
You can create a property to make access nicer e.g.
```python
@property
def transformer_layer(self):
  return getattr(f"mtp_{self.k}_transformer_layer")
```

File: src/maxtext/layers/nnx_decoders.py [hunk 11]
```diff
@@ -0,0 +1,1139 @@
...
+
+    # Workaround for Deepseek MTP test failure.
+    # TODO: Handle this properly.
+    try:
+      forked_rngs = rngs.fork(split=length)
+
+    except:  # pylint: disable=bare-except
+      pass
+
+    out_axes = nnx.StateAxes({nnx.Param: self.config.param_scan_axis, ...: 0})
```
Comment (cgarciae) [inline src/maxtext/layers/nnx_decoders.py:406]: Sorry to ask this 😅 , can we avoid using `nnx.StateAxes`? We want to deprecate it in the near future. You can use `nnx.split` and `nnx.merge` to separate the state groups e.g:

```python


def create_layer_fn(rngs):
  ...
  return nnx.split(layer, nnx.Param, ...)

graphdef, params, rest = nnx.vmap(
    create_layer_fn,
    in_axes=0,
    out_axes=(None, self.config.param_scan_axis, 0),
    axis_name="layers",
    transform_metadata={nnx.PARTITION_NAME: "layers"},
)(forked_rngs)

layers_vmapped = nnx.merge(graphdef, params, rest)
```

File: src/maxtext/layers/multi_token_prediction.py (hunk 10 above)
Comment (charlesli640) [reply src/maxtext/layers/multi_token_prediction.py:111]: Line 143-149 in this file, it should have already implemented property setter and getter as below
```
143   @property
144   def transformer_layer(self):
145     return getattr(self, f"mtp_{self.layer_number}_transformer_layer")
146 
147   @transformer_layer.setter
148   def transformer_layer(self, module):
149     setattr(self, f"mtp_{self.layer_number}_transformer_layer", module)
```

Comment (cgarciae) [reply src/maxtext/layers/multi_token_prediction.py:111]: Oh nice. Sorry I missed this.

File: src/maxtext/layers/nnx_decoders.py (hunk 11 above)
Comment (hsuan-lun-chiang) [reply src/maxtext/layers/nnx_decoders.py:406]: Thank you @cgarciae for the heads-up! I tried updating the code, but Copybara failed to migrate the changes to the CL. Since we have follow-up PRs planned for this, I'll handle the migration to `nnx.split` and `nnx.merge` there instead.
